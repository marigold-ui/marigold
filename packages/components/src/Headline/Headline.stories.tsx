import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { Headline } from './Headline';

const meta = {
  title: 'Components/Headline',
  component: Headline,
  argTypes: {
    level: {
      control: {
        type: 'select',
      },
      options: ['1', '2', '3', '4', '5', '6'],
      description: 'Level of the headline',
    },
    variant: {
      control: {
        type: 'text',
      },
      description: 'The variant of the headline',
    },
    size: {
      control: {
        type: 'text',
      },
      description: 'The size of the headline',
    },
    align: {
      control: {
        type: 'text',
      },
      description: 'Set the text-align of the component',
    },
    color: {
      control: {
        type: 'text',
      },
      description: 'Text color',
      table: {
        defaultValue: {
          summary: 'inherit',
        },
      },
    },
  },
} satisfies Meta<typeof Headline>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic: Story = {
  render: args => (
    <Headline color="red-300" {...args}>
      This is a Headline!!
    </Headline>
  ),
};
