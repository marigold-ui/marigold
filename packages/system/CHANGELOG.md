# @marigold/system

## 2.1.0

## 2.0.0

### Minor Changes

- [#2247](https://github.com/marigold-ui/marigold/pull/2247) [`8e9ea3da`](https://github.com/marigold-ui/marigold/commit/8e9ea3daaa0b3ea666394badabcfc3e76eba5512) Thanks [@sebald](https://github.com/sebald)! - feat: improve theme type

### Patch Changes

- [#2273](https://github.com/marigold-ui/marigold/pull/2273) [`4c63400f`](https://github.com/marigold-ui/marigold/commit/4c63400fef291181f29f994c7f8a8b020772cbee) Thanks [@renovate](https://github.com/apps/renovate)! - chore(deps): update dependency @emotion/react to v11.10.0

* [#2265](https://github.com/marigold-ui/marigold/pull/2265) [`88a3d4b0`](https://github.com/marigold-ui/marigold/commit/88a3d4b030e67e46a4af429b01a884195601b7a2) Thanks [@renovate](https://github.com/apps/renovate)! - chore(deps): update dependency @babel/core to v7.18.10

## 1.3.0

### Patch Changes

- [#2221](https://github.com/marigold-ui/marigold/pull/2221) [`c60f8527`](https://github.com/marigold-ui/marigold/commit/c60f8527cc4d61c3b7d8eeb2ec29a0cd7679e8dc) Thanks [@renovate](https://github.com/apps/renovate)! - chore(deps): update dependency @babel/core to v7.18.9

## 1.2.2

## 1.2.1

### Patch Changes

- [#2164](https://github.com/marigold-ui/marigold/pull/2164) [`b43464fc`](https://github.com/marigold-ui/marigold/commit/b43464fce6f7e2662b27313c6f74190e8c0f540f) Thanks [@sarahgm](https://github.com/sarahgm)! - refa: update to React 18

- Updated dependencies [[`b43464fc`](https://github.com/marigold-ui/marigold/commit/b43464fce6f7e2662b27313c6f74190e8c0f540f)]:
  - @marigold/types@0.5.2

## 1.2.0

### Minor Changes

- [#2144](https://github.com/marigold-ui/marigold/pull/2144) [`82c376a9`](https://github.com/marigold-ui/marigold/commit/82c376a9c34c69104456e6e356231132f53c3efe) Thanks [@sebald](https://github.com/sebald)! - feat: Improve usage of Marigold when not used as a full page app

## 1.1.1

## 1.1.0

## 1.0.1

### Patch Changes

- [#2102](https://github.com/marigold-ui/marigold/pull/2102) [`23c1a5ce`](https://github.com/marigold-ui/marigold/commit/23c1a5cece356465f07b7206a0fe0f65512f3350) Thanks [@sebald](https://github.com/sebald)! - feat: Remove all occurances of `React.FC`

- Updated dependencies [[`23c1a5ce`](https://github.com/marigold-ui/marigold/commit/23c1a5cece356465f07b7206a0fe0f65512f3350), [`bfa0caea`](https://github.com/marigold-ui/marigold/commit/bfa0caeac9fdb817ceed7ee30dbcce0819460f30)]:
  - @marigold/types@0.5.1

## 1.0.0

### Major Changes

- [#2074](https://github.com/marigold-ui/marigold/pull/2074) [`3aa2c100`](https://github.com/marigold-ui/marigold/commit/3aa2c100fce8884cf0e1ae8e848516923f0d8456) Thanks [@sebald](https://github.com/sebald)! - refa: Remove "variant" prop from <Box>, use `useComponentStyles` instead

### Minor Changes

- [#2041](https://github.com/marigold-ui/marigold/pull/2041) [`02d13e84`](https://github.com/marigold-ui/marigold/commit/02d13e84a27ac075587f7278c8c858aeee884f64) Thanks [@sebald](https://github.com/sebald)! - feat: Support for `:focus-visible` and `:selected` pseudos

* [#2048](https://github.com/marigold-ui/marigold/pull/2048) [`2ab80ae6`](https://github.com/marigold-ui/marigold/commit/2ab80ae677be0785cfa6e2bdf9f7ff2f3d38baec) Thanks [@sebald](https://github.com/sebald)! - refa: <Select>

- [#2063](https://github.com/marigold-ui/marigold/pull/2063) [`f5128944`](https://github.com/marigold-ui/marigold/commit/f512894439dff2b5769fe2e5aa23f61a28b3d14e) Thanks [@sebald](https://github.com/sebald)! - feat: Add shorthands for group selectors

* [#2056](https://github.com/marigold-ui/marigold/pull/2056) [`3abbc813`](https://github.com/marigold-ui/marigold/commit/3abbc8130c8b5156f3ca62f237dd59278354eb7c) Thanks [@sebald](https://github.com/sebald)! - feat: Introduce <NumberFields>

### Patch Changes

- Updated dependencies [[`02d13e84`](https://github.com/marigold-ui/marigold/commit/02d13e84a27ac075587f7278c8c858aeee884f64)]:
  - @marigold/types@0.5.0

## 1.0.0-beta.0

### Minor Changes

- [#2041](https://github.com/marigold-ui/marigold/pull/2041) [`02d13e84`](https://github.com/marigold-ui/marigold/commit/02d13e84a27ac075587f7278c8c858aeee884f64) Thanks [@sebald](https://github.com/sebald)! - feat: Support for `:focus-visible` and `:selected` pseudos

* [#2048](https://github.com/marigold-ui/marigold/pull/2048) [`2ab80ae6`](https://github.com/marigold-ui/marigold/commit/2ab80ae677be0785cfa6e2bdf9f7ff2f3d38baec) Thanks [@sebald](https://github.com/sebald)! - refa: <Select>

- [#2063](https://github.com/marigold-ui/marigold/pull/2063) [`f5128944`](https://github.com/marigold-ui/marigold/commit/f512894439dff2b5769fe2e5aa23f61a28b3d14e) Thanks [@sebald](https://github.com/sebald)! - feat: Add shorthands for group selectors

* [#2056](https://github.com/marigold-ui/marigold/pull/2056) [`3abbc813`](https://github.com/marigold-ui/marigold/commit/3abbc8130c8b5156f3ca62f237dd59278354eb7c) Thanks [@sebald](https://github.com/sebald)! - feat: Introduce <NumberFields>

### Patch Changes

- Updated dependencies [[`02d13e84`](https://github.com/marigold-ui/marigold/commit/02d13e84a27ac075587f7278c8c858aeee884f64)]:
  - @marigold/types@0.5.0-beta.0

## 0.9.0

### Patch Changes

- [#1979](https://github.com/marigold-ui/marigold/pull/1979) [`424f1705`](https://github.com/marigold-ui/marigold/commit/424f1705159d7a03dce368adadbae152acc80687) Thanks [@sebald](https://github.com/sebald)! - feat: transform state to corresponding css selector and data-attr

* [#1972](https://github.com/marigold-ui/marigold/pull/1972) [`dbc55934`](https://github.com/marigold-ui/marigold/commit/dbc55934775a6c2362525f0078ecf1595475ddc8) Thanks [@sebald](https://github.com/sebald)! - feat: introduce `useComponentStyles` hook

- [#1994](https://github.com/marigold-ui/marigold/pull/1994) [`3dff2282`](https://github.com/marigold-ui/marigold/commit/3dff2282b75ca91547f478f3305b138d1d409670) Thanks [@sarahgm](https://github.com/sarahgm)! - Refa: card-styles-structur

* [#1988](https://github.com/marigold-ui/marigold/pull/1988) [`7c1129dc`](https://github.com/marigold-ui/marigold/commit/7c1129dc140d435e5312f80a57fd48f8498c33ee) Thanks [@sarahgm](https://github.com/sarahgm)! - refa: Field Component restructure

* Updated dependencies [[`7c1129dc`](https://github.com/marigold-ui/marigold/commit/7c1129dc140d435e5312f80a57fd48f8498c33ee)]:
  - @marigold/types@0.4.1

## 0.8.0

### Minor Changes

- [#1935](https://github.com/marigold-ui/marigold/pull/1935) [`382ac6cc`](https://github.com/marigold-ui/marigold/commit/382ac6cca1865d033ba601ad269247f26ee0a13e) Thanks [@ti10le](https://github.com/ti10le)! - feat(comp): improve SVG to handle width and height prop

* [#1915](https://github.com/marigold-ui/marigold/pull/1915) [`2f45aa5d`](https://github.com/marigold-ui/marigold/commit/2f45aa5d236f861ab8187a4a574a7113d6c1f1b8) Thanks [@sebald](https://github.com/sebald)! - feat(system): theme getter

- [#1919](https://github.com/marigold-ui/marigold/pull/1919) [`5319745c`](https://github.com/marigold-ui/marigold/commit/5319745cbc20919a423dda49f4fb8ac4608009e1) Thanks [@sebald](https://github.com/sebald)! - feat: SVG supports size token from theme

## 0.7.0

## 0.6.0

### Minor Changes

- [#1817](https://github.com/marigold-ui/marigold/pull/1817) [`4cc0ad3b`](https://github.com/marigold-ui/marigold/commit/4cc0ad3b85b993e01b4d85b6f30b1a81cdee2351) Thanks [@sebald](https://github.com/sebald)! - feat(system): Add textSizeAdjust to globals

* [#1842](https://github.com/marigold-ui/marigold/pull/1842) [`e053b7b9`](https://github.com/marigold-ui/marigold/commit/e053b7b903f02c56cf10e6b9aecbedd29399895d) Thanks [@ti10le](https://github.com/ti10le)! - feat: extend svg fill prop to handle theme color

### Patch Changes

- [#1809](https://github.com/marigold-ui/marigold/pull/1809) [`539d4198`](https://github.com/marigold-ui/marigold/commit/539d41987118db125ee37a4a83231335ea15830a) Thanks [@sebald](https://github.com/sebald)! - feat(create-theme): Introduce helper package to create themes with sensible defaults

## 0.5.1

### Patch Changes

- [#1795](https://github.com/marigold-ui/marigold/pull/1795) [`a178eafe`](https://github.com/marigold-ui/marigold/commit/a178eafe8c8380ee23b4587d953ee52b231414ff) Thanks [@ti10le](https://github.com/ti10le)! - refa: use interface instead of type

## 0.5.0

### Minor Changes

- [#1723](https://github.com/marigold-ui/marigold/pull/1723) [`5936de75`](https://github.com/marigold-ui/marigold/commit/5936de75e5a0134584438117c53c5edfe15c6c5d) Thanks [@sebald](https://github.com/sebald)! - feat(system): add hook that allows to use values based on screen size

## 0.4.0

### Minor Changes

- [`f9526234`](https://github.com/marigold-ui/marigold/commit/f9526234257a149b12c14191a524691470da3942) Thanks [@sebald](https://github.com/sebald)! - chore: Use `tsup` to build packages

* [`f9526234`](https://github.com/marigold-ui/marigold/commit/f9526234257a149b12c14191a524691470da3942) Thanks [@sebald](https://github.com/sebald)! - Improved size in node_modules

### Patch Changes

- Updated dependencies [[`f9526234`](https://github.com/marigold-ui/marigold/commit/f9526234257a149b12c14191a524691470da3942)]:
  - @marigold/types@0.4.0

## 0.3.1

### Patch Changes

- [#1676](https://github.com/marigold-ui/marigold/pull/1676) [`379041bc`](https://github.com/marigold-ui/marigold/commit/379041bc7d4502bca98029a95afe43ce693222cd) Thanks [@ti10le](https://github.com/ti10le)! - remove the last mdx stories

## 0.3.0

### Minor Changes

- [#1163](https://github.com/marigold-ui/marigold/pull/1163) [`51af6693`](https://github.com/marigold-ui/marigold/commit/51af669330fd52e4e31fe5ad71d2b202ab8d2231) Thanks [@sebald](https://github.com/sebald)! - feat(system): Expose helper to transpile style object

### Patch Changes

- [#1549](https://github.com/marigold-ui/marigold/pull/1549) [`c030aa85`](https://github.com/marigold-ui/marigold/commit/c030aa85156356c294bafe9831024f6b6f2ce4db) Thanks [@sebald](https://github.com/sebald)! - refa: use emotion's theme function in <Element>

* [#1515](https://github.com/marigold-ui/marigold/pull/1515) [`8eda245f`](https://github.com/marigold-ui/marigold/commit/8eda245f01a918fcdaa9f0ac211889ed869aa375) Thanks [@sebald](https://github.com/sebald)! - feat: add normalization for body and html & fix emotion leak

- [#1491](https://github.com/marigold-ui/marigold/pull/1491) [`5a04de11`](https://github.com/marigold-ui/marigold/commit/5a04de110637d004f5824679697ee4d6a90eaf34) Thanks [@ti10le](https://github.com/ti10le)! - remove useStyles from Element

* [#1440](https://github.com/marigold-ui/marigold/pull/1440) [`c1da52c0`](https://github.com/marigold-ui/marigold/commit/c1da52c0f035b141608fd606e6ba3bc2b5482dc1) Thanks [@ti10le](https://github.com/ti10le)! - feature: use Element in Box + necessary fix changes

- [#1591](https://github.com/marigold-ui/marigold/pull/1591) [`1448ddca`](https://github.com/marigold-ui/marigold/commit/1448ddcaa0f647f48b018fa74a8686af30eccc53) Thanks [@sebald](https://github.com/sebald)! - feat(jest): Improve snapshot readability

* [#1514](https://github.com/marigold-ui/marigold/pull/1514) [`5107b943`](https://github.com/marigold-ui/marigold/commit/5107b943cb3085eb3137d84e79966acad6173a26) Thanks [@sebald](https://github.com/sebald)! - feat(system): Use emotion's context

- [#1647](https://github.com/marigold-ui/marigold/pull/1647) [`cd3a0d3e`](https://github.com/marigold-ui/marigold/commit/cd3a0d3edb3f2ddc1f561e8007e1c20000f7855a) Thanks [@ti10le](https://github.com/ti10le)! - feat: conditional function to allow state props in variants

* [#1230](https://github.com/marigold-ui/marigold/pull/1230) [`ebd6e26f`](https://github.com/marigold-ui/marigold/commit/ebd6e26f71f675b98b663bc45c6a2d5badddcd47) Thanks [@viktoria-schwarz](https://github.com/viktoria-schwarz)! - feat: add GlobalStyles via theme

- [#1563](https://github.com/marigold-ui/marigold/pull/1563) [`6e485f5a`](https://github.com/marigold-ui/marigold/commit/6e485f5a8800094fe54c075a2b21f8abe726b3cd) Thanks [@sebald](https://github.com/sebald)! - feat: Introduce a `Theme` with all available scales

* [#1620](https://github.com/marigold-ui/marigold/pull/1620) [`80a2abe5`](https://github.com/marigold-ui/marigold/commit/80a2abe5804ba2c5a48cc6b05211245c37baf266) Thanks [@ti10le](https://github.com/ti10le)! - feat(comp): extend Box to handle \_\_default variants

- [#1579](https://github.com/marigold-ui/marigold/pull/1579) [`e13e3cc1`](https://github.com/marigold-ui/marigold/commit/e13e3cc1fc66b261209973b1fc90eb48117076e9) Thanks [@ti10le](https://github.com/ti10le)! - feat: remove Heading from marigold

* [#1622](https://github.com/marigold-ui/marigold/pull/1622) [`1829cf17`](https://github.com/marigold-ui/marigold/commit/1829cf17e16c574e5577b3e1709c34dc7ed4faaf) Thanks [@ti10le](https://github.com/ti10le)! - feat(comp): change Card default variant

- [#1501](https://github.com/marigold-ui/marigold/pull/1501) [`1c1f8648`](https://github.com/marigold-ui/marigold/commit/1c1f864820a060214406ef711f4ed873746c16c4) Thanks [@ti10le](https://github.com/ti10le)! - feat(storybook): remove use styles stories

* [#1190](https://github.com/marigold-ui/marigold/pull/1190) [`a00b7eb9`](https://github.com/marigold-ui/marigold/commit/a00b7eb971131634414d3912d059fb505bb7a370) Thanks [@sebald](https://github.com/sebald)! - fix(system): List reset should not remove the enumeration

- [#1499](https://github.com/marigold-ui/marigold/pull/1499) [`ec5baf85`](https://github.com/marigold-ui/marigold/commit/ec5baf85a9a0f82a4fca5bbd1e1680316c186593) Thanks [@sebald](https://github.com/sebald)! - feat: Update and simplify normalization

* [#1652](https://github.com/marigold-ui/marigold/pull/1652) [`0bb8f19e`](https://github.com/marigold-ui/marigold/commit/0bb8f19ebdec0e2f9dc3f6164f4373cac5c10880) Thanks [@sebald](https://github.com/sebald)! - refa(system): Groupt variant related fns in own file

- [#1436](https://github.com/marigold-ui/marigold/pull/1436) [`c4ae5c5c`](https://github.com/marigold-ui/marigold/commit/c4ae5c5ca442f93034ff8f4c70adc295547951d4) Thanks [@ti10le](https://github.com/ti10le)! - create Element component + normalize file

* [#1508](https://github.com/marigold-ui/marigold/pull/1508) [`a1ef2108`](https://github.com/marigold-ui/marigold/commit/a1ef2108dd6c8e6838b517dd58c82d38e71dae2b) Thanks [@sebald](https://github.com/sebald)! - feat: Add default styling to `<Element>`

- [#1621](https://github.com/marigold-ui/marigold/pull/1621) [`2f7b936f`](https://github.com/marigold-ui/marigold/commit/2f7b936f5b07eade00a51cb138c3c492f1e08c9d) Thanks [@ti10le](https://github.com/ti10le)! - feat(comp): change Badge default variant

* [#1550](https://github.com/marigold-ui/marigold/pull/1550) [`846eb640`](https://github.com/marigold-ui/marigold/commit/846eb640ad035c7f3410b4a8a451f8de56e62339) Thanks [@sebald](https://github.com/sebald)! - feat: Merge <Box> and <Element>

- [#1614](https://github.com/marigold-ui/marigold/pull/1614) [`5d63cd9c`](https://github.com/marigold-ui/marigold/commit/5d63cd9c14578787083c82c85d93bbd2ff0efac6) Thanks [@ti10le](https://github.com/ti10le)! - remove all purple color things
