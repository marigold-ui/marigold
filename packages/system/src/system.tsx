/**
 * Typings are based on [Reach UI](https://github.com/reach/reach-ui/blob/4cb497f530b0f83f80c6f6f2da46ab55b1160cb6/packages/utils/src/types.tsx).
 */
import {
  forwardRef,
  ComponentPropsWithRef,
  ElementType,
  ReactElement,
  ValidationMap,
  WeakValidationMap,
} from 'react';

/**
 * SystemProps support the `as` and `variant` prop. The former
 * is used to changed the rendered root element of a component.
 *
 * These props also infer additional allowed props based on the
 * value of the `as` prop. For example, setting `as="button"` will
 * allow to use HTMLButtonAttributes on the component.
 */
export type SystemProps<P, T extends ElementType> = P &
  Omit<ComponentPropsWithRef<T>, 'as' | keyof P> & {
    as?: T;
    variant?: string;
  };

/**
 * Enhanced version of `React.FunctionComponent` that accepts `SystemProps`
 * and infers allowed properties based on the `as` prop.
 */
export interface SystemComponent<P, T extends ElementType> {
  /**
   * These types are a bit of a hack, but cover us in cases where the `as` prop
   * is not a JSX string type. Makes the compiler happy so ü§∑‚Äç‚ôÇÔ∏è
   */
  <TT extends ElementType>(props: SystemProps<P, TT>): ReactElement | null;
  (props: SystemProps<P, T>): ReactElement | null;

  displayName?: string;
  propTypes?: WeakValidationMap<SystemProps<P, T>>;
  contextTypes?: ValidationMap<any>;
  defaultProps?: Partial<SystemProps<P, T>>;
}

/**
 * Helper to write components that adhere to a common design system API,
 * which includes the `as` and `variant` prop.
 */
export function system<P, T extends ElementType>(
  render: (props: SystemProps<P, T>) => ReactElement | null
) {
  return forwardRef((props: any, ref) =>
    render({ ...props, ref })
  ) as SystemComponent<P, T>;
}
