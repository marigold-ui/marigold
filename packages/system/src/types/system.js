export {};
//# sourceMappingURL=system.js.map
