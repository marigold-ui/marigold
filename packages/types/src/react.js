export {};
//# sourceMappingURL=react.js.map
