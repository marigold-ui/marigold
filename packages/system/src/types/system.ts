import { Simplify } from '@marigold/types';

/**
 * Create type aliases for `theme-ui` so that it doesn't leak too much into our code.
 */
import {
  ThemeUIStyleObject,
  ThemeUICSSObject,
  ThemeUICSSProperties,
  ResponsiveStyleValue as RSV,
} from '@theme-ui/css';

export type ResponsiveStyleValue<T> = RSV<T>;
export type StyleObject = ThemeUIStyleObject;
export type CSSObject = Simplify<ThemeUICSSObject>;
export type CSSProperties = Simplify<ThemeUICSSProperties>;
