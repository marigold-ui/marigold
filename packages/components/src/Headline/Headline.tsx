import { ReactNode } from 'react';

import {
  TextAlignProp,
  cn,
  createVar,
  get,
  textAlign,
  useClassNames,
  useTheme,
} from '@marigold/system';
import { HtmlProps } from '@marigold/types';

// Props
// ---------------
export interface HeadlineProps extends HtmlProps<'h1'>, TextAlignProp {
  children?: ReactNode;
  level?: '1' | '2' | '3' | '4' | '5' | '6';
  color?: string;
  variant?: string;
  size?: string;
  as?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
}

// Component
// ---------------
export const Headline = ({
  children,
  variant,
  size,
  align = 'left',
  color,
  level = '1',
  as = `h${level}`,
  ...props
}: HeadlineProps) => {
  const theme = useTheme();
  const classNames = useClassNames({
    component: 'Headline',
    variant,
    size: size ?? `level-${level}`,
  });

  const Component = as;

  return (
    <Component
      {...props}
      className={cn(classNames, 'text-[--color]', textAlign[align])}
      style={createVar({
        color:
          color &&
          theme.colors &&
          get(theme.colors, color.replace('-', '.'), color /* fallback */),
      })}
    >
      {children}
    </Component>
  );
};
