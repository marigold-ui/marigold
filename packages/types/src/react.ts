import * as React from 'react';

/**
 * Get props from a HTML element without `ref` and strip the `style` prop.
 *
 * @example <caption>Will contain all HTML attributes of a regular button element, minus "style"</caption>
 * type ButtonProps = ComponentProps<'button'>
 */
export type HtmlProps<T extends React.ElementType> = Omit<
  React.ComponentPropsWithoutRef<T>,
  'style'
>;

/**
 * Get props from a HTML element with `ref` and strip the `style` prop.
 *
 * @example <caption>Will contain all HTML attributes of a regular button element, minus "style"</caption>
 * type ButtonProps = ComponentProps<'button'>
 */
export type HtmlPropsWithRef<T extends React.ElementType> = Omit<
  React.ComponentPropsWithRef<T>,
  'style'
>;

/**
 * Infer the props of a component `C`.
 */
export type PropsOf<C> = C extends React.FC<infer P>
  ? P
  : C extends React.Component<infer Props>
  ? Props
  : never;

/**
 * Like a regular `Omit` but handles when the passed type is `never`.
 */
export type DistributiveOmit<T, TOmitted extends PropertyKey> = T extends any
  ? Omit<T, TOmitted>
  : never;

/**
 * Fix the weird behaviro or the forwardRef types.
 * Source: https://www.totaltypescript.com/pass-component-as-prop-react
 *
 * @example const fixedForwardRef = forwardRef as FixedForwardRef;
 */
export type FixedForwardRef = <T, P = {}>(
  render: (props: P, ref: React.Ref<T>) => React.ReactNode
) => (props: P & React.RefAttributes<T>) => React.ReactNode;

/**********************************************/
/*                                            */
/*            POLYMORPIC COMPONENT            */
/*                                            */
/**********************************************/
/**
 * Polymorphic types are based on Radix's (now deprecated) types.
 * They can be found here: https://github.com/radix-ui/primitives/blob/2f139a832ba0cdfd445c937ebf63c2e79e0ef7ed/packages/react/polymorphic/src/polymorphic.ts
 */

/**
 * Merge two objects into one. Second one will override properties of the
 * first one.
 *
 * Note: We can not use `Merge` from type-fest here, because for whatever reasons
 * the optional modifier will be removed.
 *
 * @internal
 */
type Merge<P1 = {}, P2 = {}> = Omit<P1, keyof P2> & P2;

/**
 * Used to merge properties of an HTML element with additional props defined in a component.
 *
 * @internal
 */
type MergeProps<T, P = {}> = P &
  Merge<
    T extends React.ElementType
      ? Omit<React.ComponentPropsWithRef<T>, 'className'>
      : {},
    P
  >;

/**
 * Narrow down instrinsic elements to the keys (HTML tags).
 *
 * @internal
 */
type NarrowIntrinsic<E> = E extends keyof JSX.IntrinsicElements ? E : never;

/**
 * Infer the (additional) props of the component `C` if it a `PolymorphicComponent`.
 * This not include any properties added to a polymorphic component based on its
 * instrinstic element type.
 */
export type OwnProps<C> = C extends PolymorphicComponent<any, infer P>
  ? P
  : never;

export type IntrinsicElement<C> = C extends PolymorphicComponent<infer I, any>
  ? I
  : never;

export interface PolymorphicComponent<
  T,
  P = {},
  /**
   * Extends original type to ensure built in React types play nice
   * with polymorphic components still e.g. `React.ElementRef` etc.
   */
> extends React.ForwardRefExoticComponent<MergeProps<T, P & { as?: T }>> {
  /**
   * USE CASE: `as` porp is a string:
   *
   * Merges original own props (without DOM props) and the inferred props
   * from `as` element with the own props taking precendence.
   *
   * We explicitly define a `JSX.IntrinsicElements` overload so that
   * events are typed for consumers.
   */
  <As extends keyof JSX.IntrinsicElements = NarrowIntrinsic<T>>(
    props: MergeProps<As, P & { as: As }>
  ): React.ReactElement | null;

  /**
   * USE CASE: `as` porp is a component:
   *
   * Merges original own props (without DOM props) and the inferred props
   * from `as` element with the own props taking precendence.
   *
   * We don't use `React.ComponentType` here as we get type errors
   * when consumers try to do inline `as` components.
   */
  <As extends React.ElementType>(
    props: MergeProps<As, P & { as: As }>
  ): React.ReactElement | null;
}
