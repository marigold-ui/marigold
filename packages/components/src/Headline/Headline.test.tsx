import { render, screen } from '@testing-library/react';

import { Theme, ThemeProvider, cva } from '@marigold/system';

import { Headline } from './Headline';

const theme: Theme = {
  name: 'test',
  colors: {
    emerald: 'rgb(5 150 105);',
  },
  components: {
    Headline: cva('m-0 font-black', {
      variants: {
        size: {
          'level-1': 'text-[2rem]',
          'level-2': 'mb-6 text-2xl',
          'level-3': 'text-xl',
          'level-4': 'text-lg',
          'level-5': 'text-base',
          'level-6': 'text-[13px] uppercase',
        },
        variant: {
          one: 'font-small',
        },
      },
    }),
  },
};

test('renders as a "section" element', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" />
    </ThemeProvider>
  );

  const headline = screen.getByTestId('headline');
  expect(headline instanceof HTMLHeadingElement).toBeTruthy();
});

test.each(['1', '2', '3', '4', '5', '6'])(
  'uses styles based on given level from theme sizes (%s)',
  lvl => {
    render(
      <ThemeProvider theme={theme}>
        <Headline data-testid="headline" level={lvl as any} />
      </ThemeProvider>
    );

    // @ts-expect-error
    const token = theme.components.Headline({ size: `level-${lvl}` });
    expect(token).toMatchSnapshot();
  }
);

test('uses "level-1" by default', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" />
    </ThemeProvider>
  );

  const headline = screen.getByTestId('headline');

  expect(headline).toMatchInlineSnapshot(`
<h1
  class="m-0 font-black text-[2rem] text-[--color] text-left"
  data-testid="headline"
/>
`);
});

test('headline accepts a variant', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" variant="one" />
    </ThemeProvider>
  );
  const headline = screen.getByTestId('headline');
  expect(headline).toMatchInlineSnapshot(`
<h1
  class="m-0 font-black text-[2rem] font-small text-[--color] text-left"
  data-testid="headline"
/>
`);
});

test('headline accepts align property', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" align="center" />
    </ThemeProvider>
  );
  const headline = screen.getByTestId('headline');
  expect(headline).toHaveClass(`text-center`);
});

test('headline accepts other level', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" level={5} />
    </ThemeProvider>
  );
  const headline = screen.getByTestId('headline');
  expect(headline).toMatchInlineSnapshot(`
<h5
  class="m-0 font-black text-base text-[--color] text-left"
  data-testid="headline"
/>
`);
});

test('get theme color', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" color="emerald" />
    </ThemeProvider>
  );

  const headline = screen.getByTestId('headline');
  expect(headline).toMatchInlineSnapshot(`
<h1
  class="m-0 font-black text-[2rem] text-[--color] text-left"
  data-testid="headline"
  style="--color: rgb(5 150 105);;"
/>
`);
});

test('support also string as level', () => {
  render(
    <ThemeProvider theme={theme}>
      <Headline data-testid="headline" level="2" />
    </ThemeProvider>
  );

  const headline = screen.getByTestId('headline');
  expect(headline).toMatchInlineSnapshot(`
<h2
  class="m-0 font-black mb-6 text-2xl text-[--color] text-left"
  data-testid="headline"
/>
`);
});
