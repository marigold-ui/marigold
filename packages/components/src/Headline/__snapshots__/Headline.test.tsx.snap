// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses styles based on given level from theme sizes (1) 1`] = `"m-0 font-black text-[2rem]"`;

exports[`uses styles based on given level from theme sizes (2) 1`] = `"m-0 font-black mb-6 text-2xl"`;

exports[`uses styles based on given level from theme sizes (3) 1`] = `"m-0 font-black text-xl"`;

exports[`uses styles based on given level from theme sizes (4) 1`] = `"m-0 font-black text-lg"`;

exports[`uses styles based on given level from theme sizes (5) 1`] = `"m-0 font-black text-base"`;

exports[`uses styles based on given level from theme sizes (6) 1`] = `"m-0 font-black text-[13px] uppercase"`;
